lib/EarlyFunctionAnalysis/DetectABI.cpp

```diff
@@ -488,6 +494,130 @@ void DetectABI::interproceduralPropagation() {
+    const auto &Entry = Function.Entry();
+
+    auto &Summary = Oracle.getLocalFunction(Entry);
+    auto &ABI = Summary.ABIResults;
+    auto &CFG = Summary.CFG;
+    auto &WrittenRegisters = Summary.WrittenRegisters;
+
+    const bool HasNotArguments = ABI.ArgumentsRegisters.size() == 0;
+    const bool HasNotReturns = ABI.FinalReturnValuesRegisters.size() == 0;
+    const bool IsSingleNode = CFG.size() == 1;
+
+    if (HasNotReturns and HasNotArguments and IsSingleNode) {
+      auto &Block = *CFG.begin();
+
+      if (Block.Successors().size() == 1) {
+        auto &Successor = *Block.Successors().begin();
+
+        const model::TypePath *NewPrototype = nullptr;
+
+        // Select new prototype for wrapper function
+        if (auto Call = dyn_cast<efa::CallEdge>(Successor.get())) {
+          auto &Imported = Binary->ImportedDynamicFunctions();
+          auto It = Imported.find(Call->DynamicFunction());
+          if (It == Imported.end()) {
+            if (Log.isEnabled()) {
+              Log << "Could not find DynamicFunction for CallEdge in "
+                  << Block.ID().toString() << "; skipping\n";
+            }
+            continue;
+          }
+
+          const model::DynamicFunction &DynamicFunction = *It;
+          auto &Prototype = DynamicFunction.Prototype();
+
+          if (not Prototype.isValid()) {
+            if (Log.isEnabled()) {
+              Log << "Dynamic function " << DynamicFunction.OriginalName()
+                  << " has invalid prototype: " << Prototype.toString()
+                  << "; skipping\n";
+            }
+            continue;
+          }
+
+          NewPrototype = &Prototype;
+        } else if (auto ID = Successor->Destination(); ID.isValid()) {
+          using std::ranges::find_if;
+          auto ByEntryAddress =
+            [Address = ID.notInlinedAddress()](const model::Function &F) {
+              return Address == F.Entry();
+            };
+          auto Result = find_if(Binary->Functions(), ByEntryAddress);
+          revng_assert(Result != Binary->Functions().end());
+
+          NewPrototype = &(Result->Prototype());
+        }
+
+        if (NewPrototype == nullptr) {
+          continue;
+        }
+
+        using abi::FunctionType::Layout;
+        // Get layout of wrapped function
+        Layout CalleeLayout = Layout::make(*NewPrototype);
+
+        // Verify that wrapper function:
+        //  - don't write stack pointer
+        //  - don't write to callee arguments
+        //  - every store instruction writes to registers (not memory)
+        auto *BB = GCBI.getBlockAt(Block.ID().start());
+
+        auto StackPointer = [this]() {
+          const auto &Architecture = Binary->Architecture();
+          auto SP = model::Architecture::getStackPointer(Architecture);
+          llvm::StringRef StackPointerName = model::Register::getCSVName(SP);
+          return M.getGlobalVariable(StackPointerName);
+        }();
```

Comment:
Nice, but I'd avoid a lambda for a function called just once.

Reply:
I only do this trick to not make mess with temporary variable in the scope. In output code it will be optimized out, so what why should we avoid it?